src/toxcore/dht/packet/errors.rs

```diff
@@ -0,0 +1,46 @@
+/*! Error enums for DHT.
+*/
+
+use nom::{ErrorKind, Needed};
+
+use std::convert::From;
+use std::io::Error as IoError;
+use std::io::ErrorKind as IoErrorKind;
+
+/// Error that can happen when calling `get_payload` of packet.
+#[derive(Debug, Fail, PartialEq)]
+pub enum GetPayloadError {
+    /// Error indicates that received payload of encrypted packet can't be decrypted
+    #[fail(display = "Decrypt encrypted payload of {:?} packet error.", packet)]
+    DecryptError {
+        /// Packet type
+        packet: String
```

Comment:
I'd suggest to remove `packet` field. When we call `get_payload` method we know packet type statically.

Reply:
Actually I agree with Namsoo. Right now the type of the Packet is included inside ADT, and keeping a special enum may cause problems, because of inconsistency. Whereas `packet: String` is for humans and pretty printing only. But we should somehow convert it to `.context`. See https://github.com/rust-lang-nursery/failure/blob/master/examples/string_custom_error_pattern.rs#L20

Reply:
Why do we need to keep the packet type inside this ADT? When we call it inside dht server we know the packet type. So we can add this info in high level error if we need it. Using string here is tempting to put things inside that have nothing to do with packets, see: https://github.com/tox-rs/tox/pull/267#discussion_r234664423

Reply:
> Why do we need to keep the packet type inside this ADT?

I mean inside https://github.com/tox-rs/tox/blob/master/src/toxcore/dht/packet/mod.rs#L40 this ADT.

> When we call it inside dht server we know the packet type.

The question is to get the info when we are outside dht server

Reply:
So as a conclusion - we can use https://docs.rs/failure/0.1.3/failure/struct.Context.html to add a stacktrace to our errors. In this case we can check the stacktrace to see where an error appeared. So we don't need the `packet` field. Later if we need it we can add it on higher level, like:
```rust
pub enum HandlePacketError {
    ...
    GetPayloadError {
        error: GetPayloadError,
        kind: PacketKind
    },
    CookiePayloadError {
        error: GetPayloadError,
    }
    ...
}
```
Thereby, we should change `GetPayloadError` to something like:
```rust
struct GetPayloadError {
    inner: Context<GetPayloadErrorKind>
}

struct GetPayloadErrorKind {
    ... // same as current GetPayloadError, but without `packet` field
}
```

@kpp am I correct?

Reply:
Yes